### ISSUE TITLE
Detecting that the brokers are down in Producer

### ISSUE THREAD

**Issue by felipejfc** [confluentinc/confluent-kafka-go#58]
I'm trying a case where the brokers are not up and then I send a message through the producer, I expected to receive some kind of error in ```Producer.Events``` channel, however I received none.

Are there any way to detect that the brokers are not up so that I can alert that no messages are really being sent?

EDIT:
Actually after a lot of time I got an error:
```Message timed out```

Still I feel that this could be better handled...

**Comment by felipejfc** [comment]
maybe related to #27?x

**Comment by edenhill** [comment]
Brokers being unavailable is considered a temporary error and thus doesnt directly affect messages in queue waiting to be delivered, they will instead time out eventually if the appropriate leader brokers don't come back up in time.

It is generally recommended to think along these lines - constrained based delivery ("We will allow 5 minutes and 3 retries to produce a message"), rather than taking action on what is essentially technical details of the Kafka clusters runtime state, it gets complex very quickly.
Bottom line is that the application shouldn't need to worry about these things (because there is typically nothing intelligent it can do to fix things), it shall produce a message and eventually get a delivery event to propagate the result.

If your application really needs to do something when all brokers are down (or unable to connect to), then you can check for Error events with the code set to `ErrAllBrokersDown`, do note however that there currently isn't a corresponding SomeBrokersUpAgain event.

**Comment by felipejfc** [comment]
maybe this would be a good addition? (a SomeBrokersUpAgain event)

**Comment by thijsc** [comment]
Agree, it would be really useful to know when some brokers are up again.

**Comment by edenhill** [comment]
This requires underlying support in librdkafka, that feature request is tracked here:
https://github.com/edenhill/librdkafka/issues/418

**Comment by onorua** [comment]
I believe it is implemented in v0.11.6:
```
	for e := range p.Producer.Events() {
		switch ev := e.(type) {
		case *kafka.Message:
                        // reply message ....
		case kafka.Error:
			if ev.Code() == kafka.ErrAllBrokersDown {
				// do what is required when brokers are down
			}
		}
```
I believe this issue can be closed
